### PR TITLE
Use a renderer generator to create sync/async render methods with full typing

### PR DESCRIPTION
**Pull request by schloerke**
A renderer author can now supply a value that will render a value (`value_fn`) given the result of the user.

Example given `render.text`:

```python
@renderer_components
async def _text(_meta: RenderMeta, _fn: RenderFnAsync[str | None]) -> str | None:
    value = await _fn()
    if value is None:
        return None
    return str(value)


@overload
def text() -> _text.type_decorator:
    ...


@overload
def text(_fn: _text.type_renderer_fn) -> _text.type_renderer:
    ...


def text(_fn: _text.type_impl_fn = None) -> _text.type_impl:
    " DOCS "
    return _text.impl(_fn)
```

<details>
<summary>Previous definition for <code>render.text</code> (w/ docs removed for fair comparison)</summary>

```python
RenderTextFunc = Callable[[], "str | None"]
RenderTextFuncAsync = Callable[[], Awaitable["str | None"]]


class RenderText(RenderFunction["str | None", "str | None"]):
    def __init__(self, fn: RenderTextFunc) -> None:
        super().__init__(fn)
        # The Render*Async subclass will pass in an async function, but it tells the
        # static type checker that it's synchronous. wrap_async() is smart -- if is
        # passed an async function, it will not change it.
        self._fn: RenderTextFuncAsync = _utils.wrap_async(fn)

    def __call__(self) -> str | None:
        return _utils.run_coro_sync(self._run())

    async def _run(self) -> str | None:
        res = await self._fn()
        if res is None:
            return None
        return str(res)


class RenderTextAsync(RenderText, RenderFunctionAsync["str | None", "str | None"]):
    def __init__(self, fn: RenderTextFuncAsync) -> None:
        if not _utils.is_async_callable(fn):
            raise TypeError(self.__class__.__name__ + " requires an async function")
        super().__init__(typing.cast(RenderTextFunc, fn))

    async def __call__(  # pyright: ignore[reportIncompatibleMethodOverride]
        self,
    ) -> str | None:
        return await self._run()


@overload
def text(fn: RenderTextFunc | RenderTextFuncAsync) -> RenderText:
    ...


@overload
def text() -> Callable[[RenderTextFunc | RenderTextFuncAsync], RenderText]:
    ...


def text(
    fn: Optional[RenderTextFunc | RenderTextFuncAsync] = None,
) -> RenderText | Callable[[RenderTextFunc | RenderTextFuncAsync], RenderText]:
    def wrapper(fn: RenderTextFunc | RenderTextFuncAsync) -> RenderText:
        if _utils.is_async_callable(fn):
            return RenderTextAsync(fn)
        else:
            fn = typing.cast(RenderTextFunc, fn)
            return RenderText(fn)

    if fn is None:
        return wrapper
    else:
        return wrapper(fn)
```

</details>

-----------------------------------

Render methods converted: 
- [x] `text`
- [x] `plot`
- [x] `image`
- [x] `table`
- [x] `ui`

Questions
- [x] In `render.plot()`... Plot variable calls are **after** `self._fn()` has been called. Is this ok? (Previously they were **before**)
    * A: The transform function needs to take the value function and call it to get the value.
- [x] How can I update an `@overload`ed function name?
    * A: You can't. Very sad

----------------------------------

Ex renderer used in testing:

```python
from __future__ import annotations

from typing import Optional, overload

from shiny import App, Inputs, Outputs, Session, ui
from shiny.render.transformer import (
    TransformerMetadata,
    ValueFn,
    is_async_callable,
    output_transformer,
    resolve_value_fn,
)


@output_transformer
async def TestTextTransformer(
    _meta: TransformerMetadata,
    _fn: ValueFn[str | None],
    *,
    extra_txt: Optional[str] = None,
) -> str | None:
    value = await resolve_value_fn(_fn)
    value = str(value)
    value += "; "
    value += "async" if is_async_callable(_fn) else "sync"
    if extra_txt:
        value = value + "; " + str(extra_txt)
    return value


@overload
def render_test_text(
    *, extra_txt: Optional[str] = None
) -> TestTextTransformer.OutputRendererDecorator:
    ...


@overload
def render_test_text(
    _fn: TestTextTransformer.ValueFn,
) -> TestTextTransformer.OutputRenderer:
    ...


def render_test_text(
    _fn: TestTextTransformer.ValueFn | None = None,
    *,
    extra_txt: Optional[str] = None,
) -> TestTextTransformer.OutputRenderer | TestTextTransformer.OutputRendererDecorator:
    return TestTextTransformer(
        _fn,
        TestTextTransformer.params(extra_txt=extra_txt),
    )


app_ui = ui.page_fluid(
    ui.code("t1:"),
    ui.output_text_verbatim("t1"),
    ui.code("t2:"),
    ui.output_text_verbatim("t2"),
    ui.code("t3:"),
    ui.output_text_verbatim("t3"),
    ui.code("t4:"),
    ui.output_text_verbatim("t4"),
    ui.code("t5:"),
    ui.output_text_verbatim("t5"),
    ui.code("t6:"),
    ui.output_text_verbatim("t6"),
)


def server(input: Inputs, output: Outputs, session: Session):
    @output
    @render_test_text
    def t1():
        return "t1; no call"
        # return "hello"

    @output
    @render_test_text
    async def t2():
        return "t2; no call"

    @output
    @render_test_text()
    def t3():
        return "t3; call"

    @output
    @render_test_text()
    async def t4():
        return "t4; call"

    @output
    @render_test_text(extra_txt="w/ extra_txt")
    def t5():
        return "t5; call"

    @output
    @render_test_text(extra_txt="w/ extra_txt")
    async def t6():
        return "t6; call"


app = App(app_ui, server)
```
Screenshot 2023-07-13 at 4 27 15 PM

With expectations:
```python
    OutputTextVerbatim(page, "t1").expect_value("t1; no call; sync")
    OutputTextVerbatim(page, "t2").expect_value("t2; no call; async")
    OutputTextVerbatim(page, "t3").expect_value("t3; call; sync")
    OutputTextVerbatim(page, "t4").expect_value("t4; call; async")
    OutputTextVerbatim(page, "t5").expect_value("t5; call; sync; w/ extra_txt")
    OutputTextVerbatim(page, "t6").expect_value("t6; call; async; w/ extra_txt")
```

### CODE REVIEW

File: shiny/render/_render.py [hunk 1]
```diff
@@ -32,121 +24,621 @@
...
+        Returns `FALSE` as this is a synchronous renderer.
+
+    See Also
+    --------
+    * :class:`~shiny.render.Renderer`
+    * :class:`~shiny.render.RendererRun`
+    * :class:`~shiny.render.RendererAsync`
+    """
+
+    @property
```
Comment (wch) [inline shiny/render/_render.py:302]:
```suggestion
```

File: shiny/render/_render.py [hunk 2]
```diff
@@ -32,121 +24,621 @@
...
+                self.__class__.__name__ + " requires a synchronous render function"
+            )
+        # super == RendererRun
+        super().__init__(
+            render_fn,
+            handler_fn,
+            params,
+        )
+
+    def __call__(self, *_) -> OT:
```
Comment (wch) [inline shiny/render/_render.py:323]: Is the `*_` necessary?

File: shiny/render/_render.py [hunk 3]
```diff
@@ -32,121 +24,621 @@
...
+        return _utils.run_coro_sync(self._run())
```
Comment (wch) [inline shiny/render/_render.py:324]: Continue review from here.

**Review by wch — CHANGES_REQUESTED** [review]

**Review by wch — CHANGES_REQUESTED** [review]
Partial review - will continue later

File: shiny/render/_render.py [hunk 4]
```diff
@@ -32,121 +24,621 @@
...
+P = ParamSpec("P")
+
+
+# ======================================================================================
+# Helper classes
+# ======================================================================================
+
+
+# Meta information to give `hander()` some context
+class RenderMeta(NamedTuple):
```
[the remark this answers is not in the document]
Comment (schloerke) [reply shiny/render/_render.py:65]: Not for this situation. We want "immutable" information. The `NamedTuple` is defined from a `Tuple` (immutable) while `TypedDict` and dataclasses are defined using a Dict (mutable).

File: shiny/render/_render.py (hunk 2 above)
Comment (schloerke) [reply shiny/render/_render.py:323]: No. Removing

**Review by wch — CHANGES_REQUESTED** [review]

File: shiny/api-examples/output_transformer/app.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
+# Needed for types imported only during TYPE_CHECKING with Python 3.7 - 3.9
+# See https://www.python.org/dev/peps/pep-0655/#usage-in-python-3-11
```
Comment (wch) [inline shiny/api-examples/output_transformer/app.py:2]:
```suggestion
```

**Review by wch — APPROVED** [review]

File: shiny/render/_render.py [hunk 6]
```diff
@@ -2,151 +2,706 @@
...
+
+@renderer_components
+async def _text(
+    _meta: RenderMeta,
+    _fn: RenderFnAsync[str | None],
+) -> str | None:
+    value = await _fn()
+    if value is None:
+        return None
+    return str(value)
```
[the remark this answers is not in the document]
Comment (schloerke) [reply shiny/render/_render.py]: Using `ValueFn[IT]` type.  Documented when describing `transform_fn` (latest name)

**Pull request merged**